crates/typst/src/export/svg.rs

```diff
@@ -0,0 +1,604 @@
+                "id",
+                format_args!("{}{}", self.clip_paths.prefix(), id),
+            );
+            self.xml.start_element("path");
+            self.xml.write_attribute("d", &path);
+            self.xml.end_element();
+            self.xml.end_element();
+        }
+        self.xml.end_element();
+    }
+
+    /// Finalize the SVG file. This must be called after all rendering is done.
+    fn finalize(mut self) -> String {
+        self.build_clip_path();
+        self.build_glyph();
+        self.xml.end_document()
+    }
+
+    /// Render a frame with the given transform.
+    fn render_frame(&mut self, frame: &Frame, ts: Transform) {
+        self.xml.start_element("g");
+        if !ts.is_identity() {
+            self.xml.write_attribute("transform", &ts);
+        };
+        for (pos, item) in frame.items() {
+            let x = pos.x.to_pt();
+            let y = pos.y.to_pt();
+            self.xml.start_element("g");
+            self.xml
+                .write_attribute("transform", format!("translate({} {})", x, y).as_str());
+            match item {
+                FrameItem::Group(group) => self.render_group(group),
+                FrameItem::Text(text) => self.render_text(text),
+                FrameItem::Shape(shape, _) => self.render_shape(shape),
+                FrameItem::Image(image, size, _) => self.render_image(image, size),
+                FrameItem::Meta(_, _) => {}
+            };
+            self.xml.end_element();
+        }
+        self.xml.end_element();
+    }
+
+    /// Render a group. If the group has `clips` set to true, a clip path will be created.
+    fn render_group(&mut self, group: &GroupItem) {
+        self.xml.start_element("g");
+        self.xml.write_attribute("class", "typst-group");
+        if group.clips {
+            let clip_path_hash = hash128(&group).into();
+            let x = group.frame.size().x.to_pt();
+            let y = group.frame.size().y.to_pt();
+            let id = self.clip_paths.insert_with(clip_path_hash, || {
+                let mut builder = SVGPath2DBuilder(String::new());
+                builder.rect(x as f32, y as f32);
+                builder.0
+            });
+            self.xml.write_attribute_fmt(
+                "clip-path",
+                format_args!("url(#{}{})", self.clip_paths.prefix(), id),
+            );
+        }
+        self.render_frame(&group.frame, group.transform);
+        self.xml.end_element();
+    }
+
+    /// Render a text item. The text is rendered as a group of glyphs.
+    /// We will try to render the text as SVG first, then bitmap, then outline.
+    /// If none of them works, we will skip the text.
+    fn render_text(&mut self, text: &TextItem) {
+        let scale: f64 = text.size.to_pt() / text.font.units_per_em();
+        let inv_scale: f64 = text.font.units_per_em() / text.size.to_pt();
+        self.xml.start_element("g");
+        self.xml.write_attribute("class", "typst-text");
```

Comment:
Initially, I wanted to add the original text as an attribute for debugging purposes. However, it seems that xmlwriter v0.1.0 doesn't properly escape string in attributes. As a result, I finally removed it. [`b0571cb` (#1729)](https://github.com/typst/typst/pull/1729/commits/b0571cb497eb1a4d78a6dc1fbae5b8676abab150).

Related issue: https://github.com/RazrFalcon/xmlwriter/issues/9